transport/tlsfrag/record_len_writer.go

```diff
@@ -0,0 +1,332 @@
+		_, err = w.baseRF.ReadFrom(io.MultiReader(r1Reader, r2Reader))
+
+		// We should return the actual bytes read from r, not the bytes passed to base or baseRF
+		n = r1Reader.rReadLen + r2Reader.rReadLen
+		return
+	}
+	return w.baseRF.ReadFrom(r)
+}
+
+// updateSplitLen determines the split length by calling w.frag with the input of w.tlsHdr.PayloadLen().
+// It returns nil error if w.frag returns a valid split length, otherwise it returns non-nil error.
+//
+// The corresponding record lengths (including the 5 bytes header) will be set to w.r1Size and w.r2Size.
+func (w *recordLenFragWriter) updateSplitRecordLens() error {
+	if w.tlsHdr == nil {
+		return errors.New("invalid TLS header")
+	}
+	totalPayloadLen := int(w.tlsHdr.PayloadLen())
+	splitPayloadLen := w.frag(totalPayloadLen)
+	if splitPayloadLen <= 0 || splitPayloadLen >= totalPayloadLen {
+		return errors.New("invalid split position")
+	}
+	w.r1Size = splitPayloadLen + recordHeaderLen
+	w.r2Size = totalPayloadLen - splitPayloadLen + recordHeaderLen
+	return nil
+}
+
+// writeN writes at most limit bytes from p to dst.
+func writeN(dst io.Writer, p []byte, limit int) (int, error) {
+	if limit <= 0 {
+		return 0, nil
+	}
+	if len(p) > limit {
+		p = p[:limit]
+	}
+	return dst.Write(p)
+}
+
+// writeBoth writes both p1 and p2 to dst in a single Write or writev call.
+// It returns the number of bytes that are written from p1 and p2, respectively.
+//
+// Issuing a single Write or writev call to dst is required because otherwise dst
+// will receive two TCP packets, which introduces unwanted TCP split.
+//
+// Performance note, internally we might allocate a temporary buffer and copy the
+// data from p1 and p2 to that buffer, please be careful about the data size.
+func writeBoth(dst io.Writer, p1 []byte, p2 []byte) (int, int, error) {
+	var nn int64
+	var err error
+
+	if _, ok := dst.(*net.TCPConn); ok {
+		// If the underlying writer implements writev system call
+		// UDPConn and IPConn also implement writev, but TLS is TCP so we only care about TCP
+		buf := net.Buffers{p1, p2}
+		nn, err = buf.WriteTo(dst)
+	} else {
+		// We must allocate temporary buffer to hold both content and issue a single Write.
+		// This will add some pressure to GC because the temporary buffer will escape to heap.
+		// Go's proposal of memory arena can be a remedy, but the proposal is on hold indefinitely.
+		buf := bytes.NewBuffer(make([]byte, 0, len(p1)+len(p2)))
+		buf.Write(p1)
+		buf.Write(p2)
+		nn, err = buf.WriteTo(dst)
+	}
+
+	if n := int(nn); n <= len(p1) {
+		return n, 0, err
+	} else {
+		return len(p1), n - len(p1), err
+	}
+}
+
+// writeBothN writes at most limit bytes from p1 and p2 to dst in a single Write orwritev call.
+// It returns the number of bytes that are written from p1 and p2, respectively.
+func writeBothN(dst io.Writer, p1 []byte, p2 []byte, limit int) (int, int, error) {
+	if limit <= len(p1) {
+		n, err := writeN(dst, p1, limit)
```

Comment:
This would make it clearer that there's only one Write.
```suggestion
		n, err := dst.Write(p1[:limit])
```